File: Implementation/c++/22_연산자끼워넣기.cpp
// 이것이 취업을 위한 코딩테스트다 349p
// https://www.acmicpc.net/problem/14888
// 2

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
	int N;
	cin >> N;
	vector<int> A(N);
	for (int i = 0; i < N; i++)
		cin >> A[i];
	vector<int> op;
	for (int i = 0; i < 4; i++) {
		int temp;
		cin >> temp;
		for (int j = 0; j < temp; j++)
			op.push_back(i + 1);
	}
	sort(op.begin(), op.end());

	int min_value = 1e9, max_value = -1e9;
	do {
		int result = A[0];
		for (int i = 1; i < N; i++) {
			if (op[i - 1] == 1)
				result += A[i];
			else if (op[i - 1] == 2)
				result -= A[i];
			else if (op[i - 1] == 3)
				result *= A[i];
			else if (op[i - 1] == 4)
				result /= A[i];
		}
		min_value = min(min_value, result);
		max_value = max(max_value, result);
	} while (next_permutation(op.begin(), op.end()));
	cout << max_value << '\n' << min_value;
}

File: Implementation/python/22_연산자끼워넣기.py
# 이것이 취업을 위한 코딩테스트다 349p
# https://www.acmicpc.net/problem/14888
# 2

from itertools import permutations

N = int(input())
A = list(map(int, input().split()))
op_list = []
op_num = list(map(int, input().split()))
for i in range(4):
    for j in range(op_num[i]):
        op_list.append(i)

min_value = 1e9
max_value = -1e9
for op in list(permutations(op_list)):
    result = A[0]
    for i in range(1, len(A)):
        if op[i - 1] == 0:
            result += A[i]
        elif op[i - 1] == 1:
            result -= A[i]
        elif op[i - 1] == 2:
            result *= A[i]
        elif op[i - 1] == 3:
            result /= A[i]
            result = int(result)
    min_value = min(min_value, result)
    max_value = max(max_value, result)
print(max_value)
print(min_value)